### ISSUE TITLE
[fix] block-decode origin algorithm resets unmasked tokens each step

### ISSUE THREAD

**Issue by nishide-dev** [nishide-dev/unturtle#180]
## Background

The block-decode cache path (`_sample_with_cache` in `diffusion_generation_utils.py`) has a correctness bug in the `origin` algorithm that causes token match rate against the no-cache baseline to be ~41% instead of ≥90%.

### 1. Logic bug in `_sample_with_cache` origin path

At each denoising step, the origin algorithm builds a `x0` block tensor by filling it with `mask_token_id`, then only writes sampled tokens to masked positions. Tokens that were already unmasked in prior steps get reset back to `mask_token_id`:

```python
# BUG (lines ~972-980 in diffusion_generation_utils.py)
x0 = torch.full_like(mask_index_block, mask_token_id, dtype=torch.long)  # starts all-masked
x0_flat = x0[mask_index_block]
transfer = torch.rand(*x0_flat.shape, device=x.device) < p_transfer
_, sampled = sample_tokens(mask_logits[transfer], ...)
x0_flat[transfer] = sampled
x0[mask_index_block] = x0_flat
x[:, current_block_start:current_block_end] = x0  # wipes previously-unmasked tokens!
```

The no-cache path (`_sample`, origin algorithm around line 711) does this correctly — it uses `x[mask_index]` (a 1D slice of only the currently-masked positions) so unmasked tokens are never touched:

```python
# CORRECT (no-cache path)
x0 = torch.full_like(x[mask_index], mask_token_id, dtype=torch.long)
transfer = torch.rand(*x0.shape, device=x.device) < p_transfer
_, sampled = sample_tokens(mask_logits[transfer], ...)
x0[transfer] = sampled
x[mask_index] = x0  # only writes to masked positions
```

**Fix**: in the cache path, start from `x[:, current_block_start:current_block_end].clone()` rather than `torch.full_like(mask_index_block, mask_token_id)`. Then scatter sampled tokens into masked positions only.

### 2. Wrong kwarg name in `test_block_decode.py`

`test_block_decode.py` passes `block_size=8` but `MaskedDiffusionGenerationConfig` only accepts `block_length`. The unknown kwarg is silently absorbed via `setattr`, leaving `block_length=None` (single-block mode). So the tests are not actually testing multi-block decode — they pass `block_size` but exercise single-block mode.

**Fix**: replace `block_size=8` with `block_length=8` throughout `test_block_decode.py`.

## Reference

The fast-dLLM reference (`dev/repos/fast-dllm/dream/model/generation_utils_block.py`) uses a similar `x_` scratch buffer initialized from mask tokens and only scatters to masked positions via boolean index — but applies it to `x[:, current_block_start:]` in-place, leaving already-unmasked positions untouched.

## Failing tests

- `tests/models/test_block_decode.py::TestA2DBlockDecodeEquivalence::test_cache_vs_no_cache_value_equivalence` (41% match rate, expected >90%)
- `tests/models/test_llada.py::TestLLaDABlockDecode::test_block_decode_equivalence`

These tests were broken by implementation changes in commits 17eedf8b / 8326a459 after the test files were last updated (06cde890).

## Acceptance criteria

- [ ] `test_cache_vs_no_cache_value_equivalence` passes (match rate ≥90%)
- [ ] `test_block_decode_equivalence` (LLaDA) passes (exact match)
- [ ] `block_size` kwarg replaced with `block_length` in `test_block_decode.py`
- [ ] All other block decode tests continue passing
- [ ] No regression in fast tests for diffusion / models
